File: app-java/src/main/java/science/monke/internal/order/entity/OrderId.java
package science.monke.internal.order.entity;

import lombok.SneakyThrows;
import science.monke.spring.config.Environment;
import science.monke.spring.config.Region;

import javax.validation.constraints.NotNull;
import javax.validation.constraints.Null;
import java.io.Serializable;
import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;
import java.time.OffsetDateTime;
import java.util.Arrays;
import java.util.UUID;

public record OrderId(String value) implements Serializable {

  public static OrderId randomOrderId(
      @NotNull final Region region, @NotNull final Environment environment) {
    return generateOrderId(region, environment, OffsetDateTime.now(), randomSalt());
  }

  public static OrderId generateOrderId(
      @NotNull final Region region,
      @NotNull final Environment environment,
      @Null final OffsetDateTime offsetDateTime,
      @Null final String salt) {
    final byte[] bytes =
        (region.name() + environment.name() + offsetDateTime + salt)
            .getBytes(StandardCharsets.UTF_8);
    final String uuidWithoutHyphen = UUID.nameUUIDFromBytes(bytes).toString().replace("-", "");

    String identifier = "-" + region.name() + "-" + environment.name() + "-";
    if (environment == Environment.PROD) {
      identifier = "-" + region.name() + "-";
    }

    final int orderIdLength = 32;
    final int length = (orderIdLength - identifier.length()) / 2;
    String orderId =
        uuidWithoutHyphen.substring(0, length)
            + identifier
            + uuidWithoutHyphen.substring(16);
    orderId = orderId.substring(0, orderIdLength);

    return new OrderId(orderId);
  }

  @SneakyThrows
  private static String randomSalt() {
    final byte[] array = new byte[4];
    SecureRandom.getInstanceStrong().nextBytes(array);
    return Arrays.toString(array);
  }
}

File: app-java/src/main/kotlin/science/monke/internal/order/entity/OrderId.kt
package science.monke.internal.order.entity

import science.monke.spring.config.Environment
import science.monke.spring.config.Region
import java.io.Serializable
import java.nio.charset.StandardCharsets
import java.security.SecureRandom
import java.time.OffsetDateTime
import java.util.*

data class OrderId(val value: String) : Serializable {

    companion object {
        fun randomOrderId(region: Region, environment: Environment): OrderId {
            return generateOrderId(region, environment, OffsetDateTime.now(), randomSalt())
        }

        fun generateOrderId(
            region: Region,
            environment: Environment,
            offsetDateTime: OffsetDateTime,
            salt: String
        ): OrderId {
            val bytes: ByteArray =
                (region.name + environment.name + offsetDateTime + salt).toByteArray(StandardCharsets.UTF_8)
            val uuidWithoutHyphen: String = UUID.nameUUIDFromBytes(bytes).toString().replace("-", "")

            var identifier = "-$region-$environment-"
            if (environment == Environment.PROD) {
                identifier = "-$region-"
            }

            val orderIdLength = 32
            val length: Int = (orderIdLength - identifier.length) / 2
            var orderId: String = uuidWithoutHyphen.substring(0, length) + identifier + uuidWithoutHyphen.substring(16)
            orderId = orderId.substring(0, orderIdLength)

            return OrderId(orderId)
        }

        private fun randomSalt(): String {
            val byteArray = ByteArray(4)
            SecureRandom.getInstanceStrong().nextBytes(byteArray)
            return byteArray.contentToString()
        }
    }

}
